### PR TITLE
Add Youtube videos to the map

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX

#### Description

Adds youtube videos to the map. Includes a youtube view

### CODE REVIEW

**Comment by Greenheart** [comment]
@Bouncey Ah, interesting! I'll check this out locally :smile:

File: common/app/redux/fetch-youtube-saga.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+  createMapUi,
+  searchableChallengeTitles
+} from '../routes/challenges/utils';
+
+const { fetchYoutube } = types;
+export default function fetchYoutubeSaga(action$, getState) {
+  return action$
+    .filter(action => action.type === fetchYoutube)
+    .flatMap(() => {
+      const url = 'https://glitter-organisation.gomix.me/api/v1/youtube';
```
Comment (Greenheart) [inline common/app/redux/fetch-youtube-saga.js]: Should this API handle traffic from all freeCodeCamp's users? I assume it's cached for a while, but would Gomix be happy to host this API? :stuck_out_tongue:

**Review by Greenheart — DISMISSED** [review]
@Bouncey Here's some initial feedback:

1. After minimizing all sections of the map, I can't expand "Youtube" again.
2. Consider renaming the section header to "YouTube" or even "Learn Theory with freeCodeCamp's YouTube Channel" or **"Get In-depth Knowledge with freeCodeCamp's YouTube Channel"**
3. The page load time has slowed down significantly. If this is due to all the new sections and challenges that have to be added to the DOM, should we create the sections for Youtube (and possibly others "below the fold") only when campers scroll down to them? If this is caused by the API-call to get Youtube data, could it be loaded async so it doesn't block the time before the map page is starting to be usable?

Great work so far! Keep this up :smile:

File: common/app/redux/fetch-youtube-saga.js (hunk 1 above)
Comment (Bouncey) [inline common/app/redux/fetch-youtube-saga.js]: You know, I didn't even think of that...

I'll refactor that

**Comment by Bouncey** [comment]
The logic for the map UI creates a depth first search of all nodes attached to the tree, and the tree is pretty sizeable now. The PR that adds Wiki topics to the map adds something like 750 extra nodes.

Refactoring the logic behind the map is a task in itself.

The api call is of course async.

File: common/app/redux/fetch-youtube-saga.js (hunk 1 above)
Comment (Greenheart) [inline common/app/redux/fetch-youtube-saga.js]: @Bouncey Good thing we don't have to build this platform on our own :blush:

**Comment by Bouncey** [comment]
I have fixed the YouTube collapse/no-collapse issue.

**Comment by Bouncey** [comment]
I have refactored the PR to use a server-server API call which is held on the fcc server.

The fcc server:
- serves responses only from the cached response
- checks periodically if we are on a different day of the week and updates the cache accordingly.

**Comment by Greenheart** [comment]
@Bouncey Great job on the updates, I think the cached response is a good solution.

I didn't get around to verify the Collapse-issue so someone else has to look at that.

Also, I don't think I can contribute much more to this review - I'll learn more about the tech stack instead :smile:

**Comment by QuincyLarson** [comment]
@Bouncey I just tried this and it's amazing! Great work! We still need to go in and fix the styling. I'm opening a new issue specific to improving styling.

**Comment by raisedadead** [comment]
@Bouncey Could you please open the pull request once again?

**Pull request merged**